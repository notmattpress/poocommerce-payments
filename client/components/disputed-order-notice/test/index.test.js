/**
 * External dependencies
 */
import React from 'react';
import { render, screen } from '@testing-library/react';
/**
 * Internal dependencies
 */
import DisputedOrderNoticeHandler from '../index';
import { useCharge } from 'wcpay/data';

jest.mock( 'wcpay/data', () => ( {
	useCharge: jest.fn(),
} ) );

jest.mock( 'tracks', () => ( {
	recordEvent: jest.fn(),
} ) );

describe( 'DisputedOrderNoticeHandler', () => {
	const mockCharge = {
		dispute: {
			status: 'needs_response',
			reason: 'fraudulent',
			amount: 1000,
			currency: 'USD',
			evidence_details: {
				due_by: 1698500219,
			},
		},
	};

	beforeEach( () => {
		window.wcpaySettings = {
			zeroDecimalCurrencies: [],
			connect: {
				country: 'US',
			},
			dateFormat: 'M j, Y',
		};
		useCharge.mockReturnValue( { data: mockCharge } );
	} );

	afterEach( () => {
		jest.useRealTimers();
		jest.clearAllMocks();
	} );

	test( 'renders urgent dispute notice', () => {
		const fixedDate = new Date( '2023-10-28T00:00:00Z' );
		jest.useFakeTimers();
		jest.setSystemTime( fixedDate );

		const { container } = render(
			<DisputedOrderNoticeHandler
				chargeId="ch_123"
				onDisableOrderRefund={ jest.fn() }
			/>
		);
		const disputeMessages = screen.getAllByText(
			/Please resolve the dispute on this order of/
		);
		expect( disputeMessages[ 0 ] ).toBeInTheDocument();
		expect( screen.getByRole( 'button' ) ).toHaveTextContent(
			'Respond today'
		);
		expect( container ).toMatchSnapshot();
	} );

	test( 'renders regular dispute notice', () => {
		const fixedDate = new Date( '2023-10-20T00:00:00Z' );
		jest.useFakeTimers();
		jest.setSystemTime( fixedDate );

		const { container } = render(
			<DisputedOrderNoticeHandler
				chargeId="ch_123"
				onDisableOrderRefund={ jest.fn() }
			/>
		);
		const disputeMessages = screen.getAllByText( /Please respond before/ );
		expect( disputeMessages[ 0 ] ).toBeInTheDocument();
		expect( screen.getByRole( 'button' ) ).toHaveTextContent(
			'Respond now'
		);
		expect( container ).toMatchSnapshot();
	} );

	test( 'does not render notice if no dispute', () => {
		useCharge.mockReturnValue( { data: {} } );
		const { container } = render(
			<DisputedOrderNoticeHandler
				chargeId="ch_123"
				onDisableOrderRefund={ jest.fn() }
			/>
		);
		expect( container ).toBeEmptyDOMElement();
	} );

	test( 'does not render notice if dispute is not awaiting response', () => {
		mockCharge.dispute.status = 'won';
		render(
			<DisputedOrderNoticeHandler
				chargeId="ch_123"
				onDisableOrderRefund={ jest.fn() }
			/>
		);
		expect(
			screen.queryByText( /Please resolve the dispute on this order of/ )
		).not.toBeInTheDocument();
	} );
} );
