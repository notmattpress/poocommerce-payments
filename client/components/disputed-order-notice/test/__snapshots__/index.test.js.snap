// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisputedOrderNoticeHandler renders regular dispute notice 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  width: 100%;
}

.emotion-0>*+*:not( marquee ) {
  margin-left: calc(4px * 2);
}

.emotion-0>* {
  min-width: 0;
}

.emotion-2 {
  display: block;
  max-height: 100%;
  max-width: 100%;
  min-height: 0;
  min-width: 0;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  width: 100%;
}

.emotion-4>*+*:not( marquee ) {
  margin-left: calc(4px * 4);
}

.emotion-4>* {
  min-width: 0;
}

<div>
  <div
    class="wcpay-inline-notice wcpay-inline-warning-notice components-notice is-warning"
  >
    <div
      class="components-notice__content"
    >
      <div
        class="components-flex emotion-0 emotion-1"
        data-wp-c16t="true"
        data-wp-component="Flex"
      >
        <div
          class="components-flex-item wcpay-inline-notice__content wcpay-inline-warning-notice__content emotion-2 emotion-1"
          data-wp-c16t="true"
          data-wp-component="FlexItem"
        >
          <strong>
            This order has a payment dispute for $10.00 for the reason 'Transaction unauthorized'. 
          </strong>
           
          Please respond before Oct 28, 2023.
          <div
            class="components-flex wcpay-inline-notice__content__actions emotion-4 emotion-1"
            data-wp-c16t="true"
            data-wp-component="Flex"
          >
            <button
              class="components-button wcpay-inline-notice__action"
              type="button"
            >
              Respond now
            </button>
          </div>
        </div>
      </div>
      <div
        class="components-notice__actions"
      />
    </div>
  </div>
</div>
`;

exports[`DisputedOrderNoticeHandler renders urgent dispute notice 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  width: 100%;
}

.emotion-0>*+*:not( marquee ) {
  margin-left: calc(4px * 2);
}

.emotion-0>* {
  min-width: 0;
}

.emotion-2 {
  display: block;
  max-height: 100%;
  max-width: 100%;
  min-height: 0;
  min-width: 0;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  width: 100%;
}

.emotion-4>*+*:not( marquee ) {
  margin-left: calc(4px * 4);
}

.emotion-4>* {
  min-width: 0;
}

<div>
  <div
    class="wcpay-inline-notice wcpay-inline-error-notice components-notice is-error"
  >
    <div
      class="components-notice__content"
    >
      <div
        class="components-flex emotion-0 emotion-1"
        data-wp-c16t="true"
        data-wp-component="Flex"
      >
        <div
          class="components-flex-item wcpay-inline-notice__content wcpay-inline-error-notice__content emotion-2 emotion-1"
          data-wp-c16t="true"
          data-wp-component="FlexItem"
        >
          <strong>
            Please resolve the dispute on this order of $10.00 labeled 'Transaction unauthorized' by Oct 28, 2023.
          </strong>
           
          (Last day today)
          <div
            class="components-flex wcpay-inline-notice__content__actions emotion-4 emotion-1"
            data-wp-c16t="true"
            data-wp-component="Flex"
          >
            <button
              class="components-button wcpay-inline-notice__action"
              type="button"
            >
              Respond today
            </button>
          </div>
        </div>
      </div>
      <div
        class="components-notice__actions"
      />
    </div>
  </div>
</div>
`;
